README
Dezi_Client
============

Dezi_Client is a PHP client for the Dezi search platform.

Copyright 2011 by Peter Karman. Release under the MIT license. See LICENSE file.

See http://dezi.org/

# Developers ##

If you are working from a git clone of the dezi-client-php repository, you need to be aware
that the plugin uses git submodules in order to include the Dezi_Client dependencies. After
you have cloned dezi-client-php, you need to:

    % git submodule init
    % git submodule update




Changes
Revision history for dezi-client-php

0.001000  10 Oct 2011
    * initial release

0.001001  24 Oct 2011
    * fix tests to conform with Dezi server >= 0.1.5
    * set user agent on internal Pest object

0.001002  3 Sept 2012
    * support Dezi 0.001008 with commit() and rollback()

0.002000  12 Sept 2012
    * support Dezi 0.002000 with username/password basic auth

0.002001  14 Oct 2012
    * add server_supports_transactions()

0.002002  15 Oct 2012
    * set mtime in request header when indexing

0.002003  17 Oct 2012
    * add Dezi_Response->suggestions for Dezi 0.002004

1.1.1  14 Aug 2015
    * Convert to proper composer package
    * Uses proper composer naming scheme, now 1.1.1


lib/Doc.php
<?php

/*

Doc - a Dezi client document

SYNOPSIS

 // add doc to the index
 $html = "<html>hello world</html>";
 $doc = new \Dezi\Doc(array(
     'mime_type' => 'text/html',
     'uri'       => 'foo/bar.html',
     'mtime'     => time(),
     'size'      => length $html,
     'content'   => $html,
 ));
 $client->index( $doc );

 $doc2 = new \Dezi\Doc(array(
    'uri' => 'auto/xml/magic',
 ));
 $doc2->set_field('title', 'ima magic');
 $doc2->set_field('foo', array('one', 'two'));
 $doc2->set_field('body', 'hello world!');
 $client->index( $doc2 );

 // search results are also Doc objects
 foreach ($response->results as $doc) {
     printf("hit: %s %s\n", $doc->score, $doc->uri);
 }

*/

namespace Dezi;

class Doc {

    public $mime_type;
    public $summary;
    public $title;
    public $content;
    public $uri;
    public $mtime;
    public $size;
    public $score;
    private $fields = array();
    protected $VERSION = '2.0.1';

    private static $mime_types = array(
        'txt' => 'text/plain',
        'htm' => 'text/html',
        'html' => 'text/html',
        'php' => 'text/html',
        'css' => 'text/css',
        'js' => 'application/javascript',
        'json' => 'application/json',
        'xml' => 'application/xml',
        'swf' => 'application/x-shockwave-flash',
        'flv' => 'video/x-flv',

        // images
        'png' => 'image/png',
        'jpe' => 'image/jpeg',
        'jpeg' => 'image/jpeg',
        'jpg' => 'image/jpeg',
        'gif' => 'image/gif',
        'bmp' => 'image/bmp',
        'ico' => 'image/vnd.microsoft.icon',
        'tiff' => 'image/tiff',
        'tif' => 'image/tiff',
        'svg' => 'image/svg+xml',
        'svgz' => 'image/svg+xml',

        // archives
        'zip' => 'application/zip',
        'rar' => 'application/x-rar-compressed',
        'exe' => 'application/x-msdownload',
        'msi' => 'application/x-msdownload',
        'cab' => 'application/vnd.ms-cab-compressed',

        // audio/video
        'mp3' => 'audio/mpeg',
        'qt' => 'video/quicktime',
        'mov' => 'video/quicktime',

        // adobe
        'pdf' => 'application/pdf',
        'psd' => 'image/vnd.adobe.photoshop',
        'ai' => 'application/postscript',
        'eps' => 'application/postscript',
        'ps' => 'application/postscript',

        // ms office
        'doc' => 'application/msword',
        'rtf' => 'application/rtf',
        'xls' => 'application/vnd.ms-excel',
        'ppt' => 'application/vnd.ms-powerpoint',

        // open office
        'odt' => 'application/vnd.oasis.opendocument.text',
        'ods' => 'application/vnd.oasis.opendocument.spreadsheet',
    );



    /**
     *
     *
     * @param unknown $args (optional)
     */
    public function __construct($args=array()) {
        foreach ($args as $k=>$v) {
            $this->$k = $v;
        }
    }


    /**
     * Static method. Returns MIME type string for a file.
     *
     * @param string  $file
     * @return string mime_type
     */
    public static function get_mime_type($file) {
        $ext = strtolower(array_pop(explode('.', $file)));
        //error_log("file=$file ext=$ext");

        if (array_key_exists($ext, self::$mime_types)) {
            //error_log("mime=".self::$mime_types[$ext]);
            return self::$mime_types[$ext];
        }
        elseif (file_exists($file) && function_exists('mime_content_type')) {
            //error_log("mime_content_type for $file");
            return mime_content_type($file);
        }
        else {
            return 'application/octet-stream';
        }
    }


    /**
     * as_string - return Doc object, serialized.
     *
     * @return string
     */
    public function as_string() {
        if (count($this->fields)) {
            return $this->to_xml();
        }
        else {
            return $this->content;
        }
    }



    /**
     * get_field
     *
     * @param string  $name
     * @return string
     */
    public function get_field($name) {
        if (!isset($this->fields[$name])) {
            return null;
        }
        else {
            return $this->fields[$name];
        }
    }



    /**
     * set_field
     *
     * @param string  $name
     * @param string_or_array $value
     */
    public function set_field($name, $value) {
        $this->fields[$name] = $value;
    }


    /**
     * to_xml - convert fields to XML
     *
     * @return string $xml
     */
    public function to_xml() {
        $this->mime_type = 'application/xml';
        $xml = "<doc>";
        foreach ($this->fields as $f=>$v) {
            $f_safe = htmlspecialchars($f, ENT_QUOTES, 'UTF-8');
            if (is_array($v)) {
                foreach ($v as $value) {
                    $xml .= sprintf("<%s>%s</%s>", $f_safe, htmlspecialchars($value, ENT_QUOTES, 'UTF-8'), $f_safe);
                }
            }
            else {
                $xml .= sprintf("<%s>%s</%s>", $f_safe, htmlspecialchars($v, ENT_QUOTES, 'UTF-8'), $f_safe);
            }
        }
        $xml .= "</doc>";

        //error_log($xml);

        return $xml;

    }



}


lib/Client.php
<?php

/*

Client - interact with a Dezi server

SYNOPSIS

 # open a connection
 my $client = new \Dezi\Client(array(
    'server'  => 'http://localhost:5000',
 ));

 # add/update a filesystem document to the index
 $client->index( 'path/to/file.html' );

 # add/update an in-memory document to the index
 $client->index( $html_doc, 'foo/bar.html' );

 # add/update a Dezi::Doc to the index
 $client->index( $doc );

 # remove a document from the index
 $client->delete( '/doc/uri/relative/to/index' );

 # search the index
 my $response = $client->search(array( 'q' => 'foo' ));

 # iterate over results
 foreach ($response->results as $result) {
     printf("--\n uri: %s\n title: %s\n score: %s\n",
        $result->uri, $result->title, $result->score);
 }

 # print stats
 printf("       hits: %d\n", $response->total);
 printf("search time: %s\n", $response->search_time);
 printf(" build time: %s\n", $response->build_time);
 printf("      query: %s\n", $response->query);

*/

namespace Dezi;

class Client {

    public $server = 'http://localhost:5000';
    public $about_server;
    public $search_uri;
    public $index_uri;
    public $commit_uri;
    public $rollback_uri;
    public $fields;
    public $facets;
    public $last_response;
    private $username;
    private $password;
    public static $VERSION = '2.0.1';



    /**
     * Private method. Creates user agent.
     *
     * @param string  $url_base
     * @param unknown $cls      (optional)
     * @return PestJSON object
     */
    private function _new_user_agent($url_base, $cls='PestJSON') {
        $pest = new $cls($url_base);
        $pest->curl_opts[CURLOPT_USERAGENT] = 'dezi-client-php ' . self::$VERSION;
        if ($this->username && $this->password) {
            $pest->curl_opts[CURLOPT_USERPWD] = $this->username . ':' . $this->password;
        }
        return $pest;
    }


    /**
     * Constructor. Returns a new Client.
     *
     * @param array   $args (optional)
     */
    public function __construct($args=array()) {
        if (isset($args['server'])) {
            $this->server = $args['server'];
        }
        if (array_key_exists('search', $args) && array_key_exists('index', $args)) {
            $this->search_uri = $args['server'] . $args['search'];
            $this->index_uri  = $args['server'] . $args['index'];
        }
        else {

            $pest = $this->_new_user_agent($this->server);
            try {
                $resp = $pest->get('/');
            }
            catch (Pest_NotFound $err) {
                // 404
                die(sprintf("404 response from server: %s/\n", $this->server));
            }
            catch (Pest_UnknownResponse $err) {
                die(sprintf("Initial connect failed to server: %s/\n", $this->server));
            }

            //error_log( var_export($resp, true) );
            $this->about_server = $resp;
            $this->search_uri = $resp['search'];
            $this->index_uri  = $resp['index'];
            $this->fields     = $resp['fields'];
            $this->facets     = $resp['facets'];
            $this->commit_uri = $resp['commit'];
            $this->rollback_uri = $resp['rollback'];
        }
        $this->username = isset($args['username']) ? $args['username'] : null;
        $this->password = isset($args['password']) ? $args['password'] : null;

    }


    /**
     * ping() - test if the server is alive.
     *
     * @return false on failure, JSON on success
     */
    public function ping() {
        $pest = $this->_new_user_agent($this->server);
        try { $resp = $pest->get('/'); }
        catch(Pest_NotFound $err) { return false; }
        catch(Pest_UnknownResponse $err) { return false; }
        return $resp;
    }



    /**
     * index() - add or update a document in the index
     *
     * Expects either a string file, string in-memory document,
     * or a Doc object.
     *
     * @param string_or_object $doc
     * @param string  $uri          (optional)
     * @param string  $content_type (optional)
     * @return HTTPResponse $resp
     */
    public function index($doc, $uri=null, $content_type=null) {
        $buf = null;
        $mtime = time();
        if (!is_object($doc) && file_exists($doc)) {
            $buf = file_get_contents($doc);
            $mtime = filemtime($doc);
            if ($uri==null) {
                $uri = $doc;
            }
        }
        elseif (!is_object($doc)) {
            $buf = $doc;
            if ($uri==null) {
                throw new \Exception("uri required");
            }
        }
        elseif (is_object($doc) && is_a($doc, '\Dezi\Doc')) {
            $buf = $doc->as_string();
            $mtime = $doc->mtime;
            if ($uri==null) {
                $uri = $doc->uri;
            }
            if ($content_type==null) {
                $content_type = $doc->mime_type;
            }
        }
        else {
            throw new \Exception("doc must be a file, in-memory buffer, or \Dezi\Doc object");
        }

        if ($content_type==null) {
            $content_type = Doc::get_mime_type($uri);
        }

        //error_log("content_type=$content_type");
        $pest = $this->_new_user_agent($this->index_uri, 'Pest');
        $resp = $pest->post("/$uri", $buf, array(
                'Content-Type: '.$content_type,
                'X-SOS-Last-Modified: '.$mtime,
            )
        );
        $http_resp = new HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;

    }



    /**
     * search() - Fetch search results from a Dezi server.
     *
     * $params may be any key/value pair as described in Search::OpenSearch.
     *
     * Returns a Response on success and 0 on failure. Check
     * $client->last_response on failure.
     *
     * @param array   $params
     * @return Response object
     */
    public function search($params) {
        $pest = $this->_new_user_agent($this->search_uri);
        $params['t'] = 'JSON';  // force response type
        $query = http_build_query($params);
        $resp = $pest->get("?$query");
        $http_resp = new HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;
        $this->last_response = $http_resp;
        if ($pest->lastStatus() != '200') {
            return 0;
        }
        $response = new Response($resp);
        return $response;
    }



    /**
     * delete() - remove a document from the index.
     *
     * $uri should be the document URI.
     *
     * @param string  $uri
     * @return HTTPResponse $resp
     */
    public function delete($uri) {
        $pest = $this->_new_user_agent($this->index_uri, 'Pest');
        $resp = $pest->delete("/$uri");
        $http_resp = new HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;
        return $http_resp;
    }


    /**
     * Returns boolean indicating whether commit() or rollback()
     * are supported by the Dezi server.
     * Lack of server support will not prevent you from
     * calling commit() or rollback(), but you should expect
     * a 400 response from those methods if this method returns false.
     *
     * @return boolean
     */
    public function server_supports_transactions() {
        foreach ($this->about_server['methods'] as $m) {
            if ($m['description'] == 'complete any pending updates') {
                return true;
            }
        }
        return false;
    }


    /**
     * commit() - complete a transaction, saving changes to the server's index.
     *
     * @return HTTPResponse $resp
     */
    public function commit() {
        $pest = $this->_new_user_agent($this->commit_uri, 'Pest');
        $resp = $pest->post("/", "");
        $http_resp = new HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;
    }





    /**
     * rollback() - abort a transaction
     *
     * @return HTTPResponse $resp
     */
    public function rollback() {
        $pest = $this->_new_user_agent($this->rollback_uri, 'Pest');
        $resp = $pest->post("/", "");
        $http_resp = new HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;
    }


}


lib/Dezi_Doc.php
<?php

/*

Dezi_Doc - a Dezi client document

SYNOPSIS

 // add doc to the index
 require 'Dezi_Doc.php';
 $html = "<html>hello world</html>";
 $doc = new Dezi_Doc(array(
     'mime_type' => 'text/html',
     'uri'       => 'foo/bar.html',
     'mtime'     => time(),
     'size'      => length $html,
     'content'   => $html,
 ));
 $client->index( $doc );

 $doc2 = new Dezi_Doc(array(
    'uri' => 'auto/xml/magic',
 ));
 $doc2->set_field('title', 'ima magic');
 $doc2->set_field('foo', array('one', 'two'));
 $doc2->set_field('body', 'hello world!');
 $client->index( $doc2 );

 // search results are also Dezi_Doc objects
 foreach ($response->results as $doc) {
     printf("hit: %s %s\n", $doc->score, $doc->uri);
 }

*/

namespace Dezi;

class Dezi_Doc {

    public $mime_type;
    public $summary;
    public $title;
    public $content;
    public $uri;
    public $mtime;
    public $size;
    public $score;
    private $fields = array();
    protected $VERSION = '0.002003';

    private static $mime_types = array(
        'txt' => 'text/plain',
        'htm' => 'text/html',
        'html' => 'text/html',
        'php' => 'text/html',
        'css' => 'text/css',
        'js' => 'application/javascript',
        'json' => 'application/json',
        'xml' => 'application/xml',
        'swf' => 'application/x-shockwave-flash',
        'flv' => 'video/x-flv',

        // images
        'png' => 'image/png',
        'jpe' => 'image/jpeg',
        'jpeg' => 'image/jpeg',
        'jpg' => 'image/jpeg',
        'gif' => 'image/gif',
        'bmp' => 'image/bmp',
        'ico' => 'image/vnd.microsoft.icon',
        'tiff' => 'image/tiff',
        'tif' => 'image/tiff',
        'svg' => 'image/svg+xml',
        'svgz' => 'image/svg+xml',

        // archives
        'zip' => 'application/zip',
        'rar' => 'application/x-rar-compressed',
        'exe' => 'application/x-msdownload',
        'msi' => 'application/x-msdownload',
        'cab' => 'application/vnd.ms-cab-compressed',

        // audio/video
        'mp3' => 'audio/mpeg',
        'qt' => 'video/quicktime',
        'mov' => 'video/quicktime',

        // adobe
        'pdf' => 'application/pdf',
        'psd' => 'image/vnd.adobe.photoshop',
        'ai' => 'application/postscript',
        'eps' => 'application/postscript',
        'ps' => 'application/postscript',

        // ms office
        'doc' => 'application/msword',
        'rtf' => 'application/rtf',
        'xls' => 'application/vnd.ms-excel',
        'ppt' => 'application/vnd.ms-powerpoint',

        // open office
        'odt' => 'application/vnd.oasis.opendocument.text',
        'ods' => 'application/vnd.oasis.opendocument.spreadsheet',
    );



    /**
     *
     *
     * @param unknown $args (optional)
     */
    public function __construct($args=array()) {
        foreach ($args as $k=>$v) {
            $this->$k = $v;
        }
    }


    /**
     * Static method. Returns MIME type string for a file.
     *
     * @param string  $file
     * @return string mime_type
     */
    public static function get_mime_type($file) {
        $ext = strtolower(array_pop(explode('.', $file)));
        //error_log("file=$file ext=$ext");

        if (array_key_exists($ext, self::$mime_types)) {
            //error_log("mime=".self::$mime_types[$ext]);
            return self::$mime_types[$ext];
        }
        elseif (file_exists($file) && function_exists('mime_content_type')) {
            //error_log("mime_content_type for $file");
            return mime_content_type($file);
        }
        else {
            return 'application/octet-stream';
        }
    }


    /**
     * as_string - return Doc object, serialized.
     *
     * @return string
     */
    public function as_string() {
        if (count($this->fields)) {
            return $this->to_xml();
        }
        else {
            return $this->content;
        }
    }



    /**
     * get_field
     *
     * @param string  $name
     * @return string
     */
    public function get_field($name) {
        if (!isset($this->fields[$name])) {
            return null;
        }
        else {
            return $this->fields[$name];
        }
    }



    /**
     * set_field
     *
     * @param string  $name
     * @param string_or_array $value
     */
    public function set_field($name, $value) {
        $this->fields[$name] = $value;
    }


    /**
     * to_xml - convert fields to XML
     *
     * @return string $xml
     */
    public function to_xml() {
        $this->mime_type = 'application/xml';
        $xml = "<doc>";
        foreach ($this->fields as $f=>$v) {
            $f_safe = htmlspecialchars($f, ENT_QUOTES, 'UTF-8');
            if (is_array($v)) {
                foreach ($v as $value) {
                    $xml .= sprintf("<%s>%s</%s>", $f_safe, htmlspecialchars($value, ENT_QUOTES, 'UTF-8'), $f_safe);
                }
            }
            else {
                $xml .= sprintf("<%s>%s</%s>", $f_safe, htmlspecialchars($v, ENT_QUOTES, 'UTF-8'), $f_safe);
            }
        }
        $xml .= "</doc>";

        //error_log($xml);

        return $xml;

    }



}


lib/Response.php
<?php

/*

Response - Dezi search server response

SYNOPSIS

 $client = new \Dezi\Client(array(
    'server' => 'http://localhost:5000'
 ));

 $response = $client->search(array( 'q' => 'foo' ));
 // $response isa Response

 // iterate over results
 foreach ($response->results as $doc) {
     printf("--\n uri: %s\n title: %s\n score: %s\n",
        $doc->uri, $doc->title, $doc->score);
 }

 # print stats
 printf("       hits: %d\n", $response->total);
 printf("search time: %s\n", $response->search_time);
 printf(" build time: %s\n", $response->build_time);
 printf("      query: %s\n", $response->query);

*/
namespace Dezi;

class Response {

    public $results;
    public $total;
    public $search_time;
    public $build_time;
    public $query;
    public $fields;
    public $facets;
    public $suggestions;
    protected $VERSION = '2.0.1';

    /**
     * Constructor. Returns Reponse object.
     *
     * @param array   $args (optional)
     */
    public function __construct($args=array()) {
        foreach ($args as $k=>$v) {
            $this->$k = $v;
        }

        // turn results into objects
        if ($this->results) {
            $docs = array();
            $doc_fields = array();
            foreach ($this->results as $r) {
                if (!count($doc_fields)) {
                    foreach ($r as $k=>$v) {
                        if (!property_exists($k, 'Doc')) {
                            $doc_fields[] = $k;
                        }
                    }
                }
                //diag_dump($r);
                $d = new Doc($r);
                foreach ($doc_fields as $field) {
                    $d->set_field( $field, $r[$field] );
                }
                $docs[] = $d;
            }
            $this->results = $docs;
        }

    }


}


lib/Dezi_Response.php
<?php

/*

Dezi_Response - Dezi search server response

SYNOPSIS

 require 'Dezi_Response.php';
 $client = new Dezi_Client(array(
    'server' => 'http://localhost:5000'
 ));

 $response = $client->search(array( 'q' => 'foo' ));
 // $response isa Dezi_Response

 // iterate over results
 foreach ($response->results as $doc) {
     printf("--\n uri: %s\n title: %s\n score: %s\n",
        $doc->uri, $doc->title, $doc->score);
 }

 # print stats
 printf("       hits: %d\n", $response->total);
 printf("search time: %s\n", $response->search_time);
 printf(" build time: %s\n", $response->build_time);
 printf("      query: %s\n", $response->query);

*/
namespace Dezi;

class Dezi_Response {

    public $results;
    public $total;
    public $search_time;
    public $build_time;
    public $query;
    public $fields;
    public $facets;
    public $suggestions;
    protected $VERSION = '0.002003';

    /**
     * Constructor. Returns Dezi_Reponse object.
     *
     * @param array   $args (optional)
     */
    public function __construct($args=array()) {
        foreach ($args as $k=>$v) {
            $this->$k = $v;
        }

        // turn results into objects
        if ($this->results) {
            $docs = array();
            $doc_fields = array();
            foreach ($this->results as $r) {
                if (!count($doc_fields)) {
                    foreach ($r as $k=>$v) {
                        if (!property_exists($k, 'Dezi_Doc')) {
                            $doc_fields[] = $k;
                        }
                    }
                }
                //diag_dump($r);
                $d = new Dezi_Doc($r);
                foreach ($doc_fields as $field) {
                    $d->set_field( $field, $r[$field] );
                }
                $docs[] = $d;
            }
            $this->results = $docs;
        }

    }


}


lib/Dezi_Client.php
<?php


// $pest_path = realpath(dirname(__FILE__).'/../pest');
// //error_log("pest_path=$pest_path");
// set_include_path(get_include_path() . PATH_SEPARATOR . $pest_path);
// require_once 'PestJSON.php';
// require_once 'Dezi_Doc.php';
// require_once 'Dezi_Response.php';
// require_once 'Dezi_HTTPResponse.php';

/*

Dezi_Client - interact with a Dezi server

SYNOPSIS

 require_once 'Dezi_Client.php';

 # open a connection
 my $client = new Dezi::Client(array(
    'server'  => 'http://localhost:5000',
 ));

 # add/update a filesystem document to the index
 $client->index( 'path/to/file.html' );

 # add/update an in-memory document to the index
 $client->index( $html_doc, 'foo/bar.html' );

 # add/update a Dezi::Doc to the index
 $client->index( $dezi_doc );

 # remove a document from the index
 $client->delete( '/doc/uri/relative/to/index' );

 # search the index
 my $response = $client->search(array( 'q' => 'foo' ));

 # iterate over results
 foreach ($response->results as $result) {
     printf("--\n uri: %s\n title: %s\n score: %s\n",
        $result->uri, $result->title, $result->score);
 }

 # print stats
 printf("       hits: %d\n", $response->total);
 printf("search time: %s\n", $response->search_time);
 printf(" build time: %s\n", $response->build_time);
 printf("      query: %s\n", $response->query);

*/

namespace Dezi;

class Dezi_Client {

    public $server = 'http://localhost:5000';
    public $about_server;
    public $search_uri;
    public $index_uri;
    public $commit_uri;
    public $rollback_uri;
    public $fields;
    public $facets;
    public $last_response;
    private $username;
    private $password;
    public static $VERSION = '0.002003';



    /**
     * Private method. Creates user agent.
     *
     * @param string  $url_base
     * @param unknown $cls      (optional)
     * @return PestJSON object
     */
    private function _new_user_agent($url_base, $cls='PestJSON') {
        $pest = new $cls($url_base);
        $pest->curl_opts[CURLOPT_USERAGENT] = 'dezi-client-php ' . self::$VERSION;
        if ($this->username && $this->password) {
            $pest->curl_opts[CURLOPT_USERPWD] = $this->username . ':' . $this->password;
        }
        return $pest;
    }


    /**
     * Constructor. Returns a new Client.
     *
     * @param array   $args (optional)
     */
    public function __construct($args=array()) {
        if (isset($args['server'])) {
            $this->server = $args['server'];
        }
        if (array_key_exists('search', $args) && array_key_exists('index', $args)) {
            $this->search_uri = $args['server'] . $args['search'];
            $this->index_uri  = $args['server'] . $args['index'];
        }
        else {

            $pest = $this->_new_user_agent($this->server);
            try {
                $resp = $pest->get('/');
            }
            catch (Pest_NotFound $err) {
                // 404
                die(sprintf("404 response from server: %s/\n", $this->server));
            }
            catch (Pest_UnknownResponse $err) {
                die(sprintf("Initial connect failed to server: %s/\n", $this->server));
            }

            //error_log( var_export($resp, true) );
            $this->about_server = $resp;
            $this->search_uri = $resp['search'];
            $this->index_uri  = $resp['index'];
            $this->fields     = $resp['fields'];
            $this->facets     = $resp['facets'];
            $this->commit_uri = $resp['commit'];
            $this->rollback_uri = $resp['rollback'];
        }
        $this->username = isset($args['username']) ? $args['username'] : null;
        $this->password = isset($args['password']) ? $args['password'] : null;

    }


    /**
     * ping() - test if the server is alive.
     *
     * @return false on failure, JSON on success
     */
    public function ping() {
        $pest = $this->_new_user_agent($this->server);
        try { $resp = $pest->get('/'); }
        catch(Pest_NotFound $err) { return false; }
        catch(Pest_UnknownResponse $err) { return false; }
        return $resp;
    }



    /**
     * index() - add or update a document in the index
     *
     * Expects either a string file, string in-memory document,
     * or a Dezi_Doc object.
     *
     * @param string_or_object $doc
     * @param string  $uri          (optional)
     * @param string  $content_type (optional)
     * @return Dezi_HTTPResponse $resp
     */
    public function index($doc, $uri=null, $content_type=null) {
        $buf = null;
        $mtime = time();
        if (!is_object($doc) && file_exists($doc)) {
            $buf = file_get_contents($doc);
            $mtime = filemtime($doc);
            if ($uri==null) {
                $uri = $doc;
            }
        }
        elseif (!is_object($doc)) {
            $buf = $doc;
            if ($uri==null) {
                throw new \Exception("uri required");
            }
        }
        elseif (is_object($doc) && is_a($doc, '\Dezi\Dezi_Doc')) {
            $buf = $doc->as_string();
            $mtime = $doc->mtime;
            if ($uri==null) {
                $uri = $doc->uri;
            }
            if ($content_type==null) {
                $content_type = $doc->mime_type;
            }
        }
        else {
            throw new \Exception("doc must be a file, in-memory buffer, or \Dezi\Dezi_Doc object");
        }

        if ($content_type==null) {
            $content_type = Dezi_Doc::get_mime_type($uri);
        }

        //error_log("content_type=$content_type");
        $pest = $this->_new_user_agent($this->index_uri, 'Pest');
        $resp = $pest->post("/$uri", $buf, array(
                'Content-Type: '.$content_type,
                'X-SOS-Last-Modified: '.$mtime,
            )
        );
        $http_resp = new Dezi_HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;

    }



    /**
     * search() - Fetch search results from a Dezi server.
     *
     * $params may be any key/value pair as described in Search::OpenSearch.
     *
     * Returns a Dezi_Response on success and 0 on failure. Check
     * $client->last_response on failure.
     *
     * @param array   $params
     * @return Dezi_Response object
     */
    public function search($params) {
        $pest = $this->_new_user_agent($this->search_uri);
        $params['t'] = 'JSON';  // force response type
        $query = http_build_query($params);
        $resp = $pest->get("?$query");
        $http_resp = new Dezi_HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;
        $this->last_response = $http_resp;
        if ($pest->lastStatus() != '200') {
            return 0;
        }
        $dezi_response = new Dezi_Response($resp);
        return $dezi_response;
    }



    /**
     * delete() - remove a document from the index.
     *
     * $uri should be the document URI.
     *
     * @param string  $uri
     * @return Dezi_HTTPResponse $resp
     */
    public function delete($uri) {
        $pest = $this->_new_user_agent($this->index_uri, 'Pest');
        $resp = $pest->delete("/$uri");
        $http_resp = new Dezi_HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;
        return $http_resp;
    }


    /**
     * Returns boolean indicating whether commit() or rollback()
     * are supported by the Dezi server.
     * Lack of server support will not prevent you from
     * calling commit() or rollback(), but you should expect
     * a 400 response from those methods if this method returns false.
     *
     * @return boolean
     */
    public function server_supports_transactions() {
        foreach ($this->about_server['methods'] as $m) {
            if ($m['description'] == 'complete any pending updates') {
                return true;
            }
        }
        return false;
    }


    /**
     * commit() - complete a transaction, saving changes to the server's index.
     *
     * @return Dezi_HTTPResponse $resp
     */
    public function commit() {
        $pest = $this->_new_user_agent($this->commit_uri, 'Pest');
        $resp = $pest->post("/", "");
        $http_resp = new Dezi_HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;
    }





    /**
     * rollback() - abort a transaction
     *
     * @return Dezi_HTTPResponse $resp
     */
    public function rollback() {
        $pest = $this->_new_user_agent($this->rollback_uri, 'Pest');
        $resp = $pest->post("/", "");
        $http_resp = new Dezi_HTTPResponse();
        $http_resp->status = $pest->lastStatus();
        $http_resp->content = $resp;

        return $http_resp;
    }


}
